examples_apps/run_state_learner.py

```diff
@@ -0,0 +1,453 @@
+# Apply circuit of layers with corresponding depth
+with prog.context as q:
+    for k in range(depth):
+        layer(k) | q[0]
+
+
+######################################################################
+# Performing the optimization
+# ---------------------------
+#
+# :math:`\newcommand{ket}[1]{\left|#1\right\rangle}` With the Strawberry
+# Fields TensorFlow backend calculating the resulting state of the circuit
+# symbolically, we can use TensorFlow to optimize the gate parameters to
+# minimize the cost function we specify. With state learning, the measure
+# of distance between two quantum states is given by the fidelity of the
+# output state :math:`\ket{\psi}` with some target state
+# :math:`\ket{\psi_t}`. This is defined as the overlap between the two
+# states:
+#
+# .. math::  F = \left|\left\langle{\psi}\mid{\psi_t}\right\rangle\right|^2
+#
+# where the output state can be written
+# :math:`\ket{\psi}=U(\vec{\theta})\ket{\psi_0}`, with
+# :math:`U(\vec{\theta})` the unitary operation applied by the variational
+# quantum circuit, and :math:`\ket{\psi_0}=\ket{0}` the initial state.
+#
+# Let's define the target state as the single photon state
+# :math:`\ket{\psi_t}=\ket{1}`:
+
+import numpy as np
+
+target_state = np.zeros([cutoff])
+target_state[1] = 1
+print(target_state)
+
+
+######################################################################
+# Using this target state, we calculate the fidelity with the state
+# exiting the variational circuit. We must use TensorFlow functions to
+# manipulate this data, as well as a ``GradientTape`` to keep track of the
+# corresponding gradients!
+#
+# We choose the following cost function:
+#
+# .. math:: C(\vec{\theta}) = \left| \langle \psi_t \mid U(\vec{\theta})\mid 0\rangle - 1\right|
+#
+# By minimizing this cost function, the variational quantum circuit will
+# prepare a state with high fidelity to the target state.
+
+
+def cost(weights):
+    # Create a dictionary mapping from the names of the Strawberry Fields
+    # free parameters to the TensorFlow weight values.
+    mapping = {p.name: w for p, w in zip(sf_params.flatten(), tf.reshape(weights, [-1]))}
+
+    # Run engine
+    state = eng.run(prog, args=mapping).state
+
+    # Extract the statevector
+    ket = state.ket()
+
+    # Compute the fidelity between the output statevector
+    # and the target state.
+    fidelity = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state)) ** 2
+
+    # Objective function to minimize
+    cost = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state) - 1)
+    return cost, fidelity, ket
+
+
+#######################################################################
+# Now that the cost function is defined, we can define and run the
+# optimization. Below, we choose the Adam
+# optimizer that is built into TensorFlow, and instantiate
+# the Strawberry Fields TensorFlow backend.
+
+opt = tf.keras.optimizers.Adam(learning_rate=lr)
+eng = sf.Engine("tf", backend_options={"cutoff_dim": cutoff})
```

Comment:
`eng` is used in the cost function, would it make sense to define it earlier?

Reply:
Good catch, I'll move it earlier!